src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/FileServerHandler.java

```diff
@@ -65,20 +66,34 @@ public final class FileServerHandler implements HttpHandler {
     private final Logger logger;
 
     private FileServerHandler(Path root, UnaryOperator<String> mimeTable) {
-        root = root.normalize();
-        if (!Files.exists(root))
-            throw new IllegalArgumentException("Path does not exist: " + root);
-        if (!root.isAbsolute())
-            throw new IllegalArgumentException("Path is not absolute: " + root);
-        if (!Files.isDirectory(root))
-            throw new IllegalArgumentException("Path is not a directory: " + root);
-        if (!Files.isReadable(root))
-            throw new IllegalArgumentException("Path is not readable: " + root);
-        this.root = root;
+        this.root = readRoot(root);
         this.mimeTable = mimeTable;
         this.logger = System.getLogger("com.sun.net.httpserver");
     }
 
+    private static Path readRoot(Path root) {
+        // `toRealPath()` invocation below already checks if file exists, though
+        // there is no way to figure out if it fails due to a non-existent file.
+        // Hence, checking the existence here first to deliver the user a more
+        // descriptive message.
+        if (!Files.exists(root)) {
+            throw new IllegalArgumentException("Path does not exist: " + root);
+        }
```

Comment:
As I tried to explain in the comment block above, without this pre-check, non-existent files are reported poorly:

```
$ build/linux-x64/jdk/bin/jwebserver -d no-such-file
Error: server config failed: Path is invalid: no-such-file
$ _
```

Note the lack of context regarding the failure.